### PR TITLE
Issue #627 Stripping write-behind back to the bone

### PR DESCRIPTION
**Pull request by chrisdennis**
This commit removes two features of write behind (and cleans up a needlessly exposed concept) with the proposal that both these features (retry and rate limiting) can be replaced with `CacheLoaderWriter` wrappers dedicated to this task.  This simplifies the central logic of write behind.

In this initial PR I have not added implementations of these wrappers.  I would be interested in getting feedback on whether we think these should be provided by us now (or ever).

### CODE REVIEW

**Comment by ljacomet** [comment]
Ok with the removal, we should consider providing the wrapper implementations although later.

File: impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java [hunk 1]
```diff
@@ -519,28 +467,10 @@ private void processSingleOperation(List<SingleOperation<K, V>> quarantinedItems
...
-            } catch (InterruptedException e1) {
-              Thread.currentThread().interrupt();
-              throw new Exception("Exception while processing key '" + item.getKey() + "' write behind queue", e);
-            }
-          }
-        }
+      try {
+        item.performSingleOperation(cacheLoaderWriter);
+      } catch (Exception e) {
+        LOGGER.warn("Exception while processing key '{}' write behind queue : {}", item.getKey(), e);
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java]: I believe we need to delegate to the resilience strategy here and by default remove the mapping from the cache.

File: impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java [hunk 2]
```diff
@@ -398,60 +381,25 @@ private void processBatchedOperations(List<SingleOperation<K, V>> quarantinedIte
...
+        for (SingleOperation<K, V> singleOperation : batch) {
+          if (successes.contains(singleOperation.getKey()))
+            batch.remove(singleOperation);
+        }
+        if (failures != null) {
+          for (Map.Entry<?, Exception> entry : failures.entrySet()) {
+            LOGGER.warn("Exception while processing key '{}' write behind queue", entry.getKey());
           }
         }
+      } catch (Exception e) {
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java]: I believe we need to delegate to the resilience strategy here and by default remove the mappings from the cache

File: impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java [hunk 3]
```diff
@@ -398,60 +381,25 @@ private void processBatchedOperations(List<SingleOperation<K, V>> quarantinedIte
...
+        Map<?, Exception> failures = bulkCacheWritingException.getFailures();
+        Set<?> successes = bulkCacheWritingException.getSuccesses();
+        // remove successful items
+        for (SingleOperation<K, V> singleOperation : batch) {
+          if (successes.contains(singleOperation.getKey()))
+            batch.remove(singleOperation);
+        }
+        if (failures != null) {
+          for (Map.Entry<?, Exception> entry : failures.entrySet()) {
+            LOGGER.warn("Exception while processing key '{}' write behind queue", entry.getKey());
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java]: We should delegate to the resilience strategy and remove the set of failed keys from the cache.

**Comment by ljacomet** [comment]
The run time of the write behind tests is not acceptable. They add 30 seconds to the build. This was introduced in the PR for #626 and should be fixed here.

File: impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java (hunk 3 above)
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java]: I differ slightly in the details of how this should be handled, but regardless I had intended to fix this under #628

**Pull request merged**